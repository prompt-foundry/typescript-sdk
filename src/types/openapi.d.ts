import type {
  OpenAPIClient,
  Parameters,
  UnknownParamsObject,
  OperationResponse,
  AxiosRequestConfig,
} from 'openapi-client-axios';

declare namespace Components {
    namespace Schemas {
        export interface Error {
            /**
             * example:
             * Prompt ID is required.
             */
            error?: string;
        }
        export interface PromptConfiguration {
            /**
             * example:
             * PROMPT_1
             */
            promptId: string;
            promptParameters: {
                /**
                 * example:
                 * 0
                 */
                frequencyPenalty: number;
                /**
                 * example:
                 * gpt-3.5-turbo
                 */
                modelName: string;
                /**
                 * example:
                 * 0
                 */
                presencePenalty: number;
                /**
                 * example:
                 * text
                 */
                responseFormat: "text" | "json";
                /**
                 * example:
                 * 1
                 */
                temperature: number;
                /**
                 * example:
                 * 1
                 */
                topP: number;
                /**
                 * example:
                 * 100
                 */
                maxTokens: number | null;
                /**
                 * example:
                 * 97946543
                 */
                seed: number | null;
            };
            /**
             * Variables to be used in the prompt messages.
             * example:
             * [
             *   "name",
             *   "city"
             * ]
             */
            promptVariables: string[];
            promptMessages: {
                /**
                 * example:
                 * Hello, {{city}}!
                 */
                content: string;
                /**
                 * example:
                 * user
                 */
                role: "USER" | "ASSISTANT" | "SYSTEM" | "TOOL";
            }[];
        }
    }
}
declare namespace Paths {
    namespace SdkV1Prompts$PromptId {
        namespace Get {
            export interface HeaderParameters {
                "X-API-KEY": Parameters.XAPIKEY;
            }
            namespace Parameters {
                export type PromptId = string;
                export type XAPIKEY = string;
            }
            export interface PathParameters {
                promptId: Parameters.PromptId;
            }
            namespace Responses {
                export type $200 = Components.Schemas.PromptConfiguration;
                export type $400 = Components.Schemas.Error;
                export type $401 = Components.Schemas.Error;
                export type $404 = Components.Schemas.Error;
                export type $500 = Components.Schemas.Error;
            }
        }
    }
}

export interface OperationMethods {
}

export interface PathsDictionary {
  ['/sdk/v1/prompts/{promptId}']: {
  }
}

export type Client = OpenAPIClient<OperationMethods, PathsDictionary>
