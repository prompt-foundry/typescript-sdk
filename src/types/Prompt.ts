export type responseFormat = {
  type: 'text' | 'json_object'
}

export type PromptParameters = {
  frequencyPenalty: number
  maxTokens?: number
  modelName: string
  presencePenalty: number
  responseFormat: responseFormat
  seed?: number
  temperature: number
  topP: number
}

export enum PromptMessageRole {
  USER = 'user',
  ASSISTANT = 'assistant',
  TOOL = 'tool',
  SYSTEM = 'system'
}

export type PromptMessage = {
  content: string
  role: PromptMessageRole
}

export type Prompt = {
  promptId: string
  promptParameters: PromptParameters
  promptMessages: PromptMessage[]
}
