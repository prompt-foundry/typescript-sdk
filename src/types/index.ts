import { Components } from './openapi'

export type PromptConfiguration = Components.Schemas.PromptConfiguration
export type PromptMessage = Components.Schemas.PromptConfiguration['promptMessages'][0]
