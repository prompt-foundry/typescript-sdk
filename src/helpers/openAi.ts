import {
  ChatCompletionCreateParamsNonStreaming,
  ChatCompletionMessageParam,
  ChatCompletionMessageToolCall,
  ChatCompletionRole,
  ChatCompletionTool,
  ChatCompletionToolChoiceOption
} from 'openai/resources'

import { PromptConfiguration, PromptMessage, PromptTool } from '../types'

export const mapMessagesToOpenAI = (messages: PromptMessage[]): ChatCompletionMessageParam[] => {
  return messages.map((message): ChatCompletionMessageParam => {
    const role = message.role.toLowerCase() as ChatCompletionRole
    if (role === 'tool') {
      if (!message.toolCallId) {
        throw new Error('Tool call missing tool call id')
      }

      if (!message.content) {
        throw new Error('Tool message missing content')
      }

      return {
        role,
        content: message.content,
        tool_call_id: message.toolCallId
      }
    }
    if (role === 'assistant') {
      return {
        role,
        content: message.content,
        name: undefined,
        tool_calls: message.toolCalls?.map((toolCall): ChatCompletionMessageToolCall => {
          return {
            id: toolCall.toolCallId,
            type: toolCall.type,
            function: toolCall.function
          }
        })
      }
    }

    if (role === 'user') {
      if (!message.content) {
        throw new Error('User message missing content')
      }

      return {
        role,
        name: undefined,
        content: message.content
      }
    }

    if (role === 'system') {
      if (!message.content) {
        throw new Error('System message missing content')
      }

      return {
        role,
        content: message.content
      }
    }

    throw new Error(`Invalid message role: ${role}`)
  })
}

export const mapToolChoiceToOpenAI = (tools: PromptTool[], toolChoice?: string | null): ChatCompletionToolChoiceOption | undefined => {
  if (tools.length === 0) {
    return undefined
  }
  if (toolChoice === 'auto' || (!toolChoice && tools.length !== 0)) {
    return 'auto'
  }
  if (toolChoice === 'none') {
    return 'none'
  }

  if (toolChoice) {
    const tool = tools.find((t) => t.name === toolChoice)
    if (tool) {
      return {
        type: 'function',
        function: {
          name: tool.name
        }
      }
    }
  }

  return 'none'
}

const mapToolToOpenAi = (tool: PromptTool): ChatCompletionTool => {
  return {
    type: 'function',
    function: {
      name: tool.name,
      description: tool.description,
      parameters: tool.parameters
    }
  }
}

const getTools = (tools: PromptTool[]): ChatCompletionTool[] | undefined => {
  if (tools.length === 0) {
    return undefined
  }

  return tools.map((tool) => mapToolToOpenAi(tool))
}

export const mapPromptToOpenAIConfig = (promptConfig: PromptConfiguration): ChatCompletionCreateParamsNonStreaming => {
  const { messages: promptMessages, parameters, tools } = promptConfig

  const messages = mapMessagesToOpenAI(promptMessages)

  return {
    messages,
    model: parameters.modelName,
    top_p: parameters.topP,
    max_tokens: parameters.maxTokens,
    temperature: parameters.temperature,
    seed: parameters.seed,
    presence_penalty: parameters.presencePenalty,
    frequency_penalty: parameters.frequencyPenalty,
    tool_choice: mapToolChoiceToOpenAI(tools, parameters.toolChoice),
    response_format: {
      type: parameters.responseFormat === 'JSON' ? 'json_object' : 'text'
    },
    tools: getTools(tools)
  }
}
