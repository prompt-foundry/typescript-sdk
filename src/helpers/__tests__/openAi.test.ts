import { createPromptConfigurationFixture } from '../../test/__fixtures__/promptConfiguration'
import { createPromptMessageFixture } from '../../test/__fixtures__/promptMessage'
import { createPromptToolFixture } from '../../test/__fixtures__/promptTool'
import { PromptMessageRoleEnum, PromptTool } from '../../types'
import { mapMessagesToOpenAI, mapPromptToOpenAIConfig, mapToolChoiceToOpenAI } from '../openAi'

describe('openAi helpers', () => {
  describe('mapMessagesToOpenAI', () => {
    const mockMessages = [
      createPromptMessageFixture({ role: PromptMessageRoleEnum.TOOL, content: 'content1', toolCallId: 'id1' }),
      createPromptMessageFixture({ role: PromptMessageRoleEnum.ASSISTANT, content: 'content2' }),
      createPromptMessageFixture({ role: PromptMessageRoleEnum.USER, content: 'content3' }),
      createPromptMessageFixture({ role: PromptMessageRoleEnum.SYSTEM, content: 'content4' })
    ]

    it('should correctly map tool messages', () => {
      const result = mapMessagesToOpenAI([mockMessages[0]])
      expect(result[0]).toEqual({
        role: 'tool',
        content: 'content1',
        tool_call_id: 'id1'
      })
    })

    it('should correctly map assistant messages', () => {
      const result = mapMessagesToOpenAI([mockMessages[1]])
      expect(result[0]).toEqual({
        role: 'assistant',
        content: 'content2',
        name: undefined,
        tool_calls: undefined
      })
    })

    it('should correctly map user messages', () => {
      const result = mapMessagesToOpenAI([mockMessages[2]])
      expect(result[0]).toEqual({
        role: 'user',
        name: undefined,
        content: 'content3'
      })
    })

    it('should correctly map system messages', () => {
      const result = mapMessagesToOpenAI([mockMessages[3]])
      expect(result[0]).toEqual({
        role: 'system',
        content: 'content4'
      })
    })

    it('should throw an error for invalid message roles', () => {
      expect(() => mapMessagesToOpenAI([createPromptMessageFixture({ role: 'invalid', content: 'content5' } as any)])).toThrow(
        'Invalid message role: invalid'
      )
    })

    // Additional test cases for missing fields and other edge cases can be added here.
  })

  describe('mapPromptToOpenAIConfig', () => {
    it('should map the configuration to OpenAI parameters correctly', () => {
      const result = mapPromptToOpenAIConfig(createPromptConfigurationFixture())
      expect(result).toMatchSnapshot()
    })

    it('should map the configuration to OpenAI parameters correctly - tools', () => {
      const result = mapPromptToOpenAIConfig(
        createPromptConfigurationFixture({
          promptTools: [createPromptToolFixture({ name: 'exampleTool' })]
        })
      )
      expect(result).toMatchSnapshot()
    })

    // Additional test cases to test other configurations and scenarios can be added here.
  })

  describe('mapToolChoiceToOpenAI', () => {
    it('should return "auto" if toolChoice is "auto" or there are tools available and toolChoice is undefined', () => {
      const tools = [createPromptToolFixture()]
      expect(mapToolChoiceToOpenAI(tools, 'auto')).toBe('auto')
      expect(mapToolChoiceToOpenAI(tools)).toBe('auto')
    })

    it('should return "none" if there are no tools or toolChoice is "none"', () => {
      const tools: PromptTool[] = []
      expect(mapToolChoiceToOpenAI(tools, 'none')).toBe(undefined)
      expect(mapToolChoiceToOpenAI(tools)).toBe(undefined)
      expect(mapToolChoiceToOpenAI(tools, 'auto')).toBe(undefined)
    })

    it('should return tool function object if a valid toolChoice matches a tool', () => {
      const tools = [createPromptToolFixture({ name: 'exampleTool' })]
      expect(mapToolChoiceToOpenAI(tools, 'exampleTool')).toEqual({
        type: 'function',
        function: { name: 'exampleTool' }
      })
    })

    it('should return "none" if the toolChoice does not match any tool', () => {
      const tools = [createPromptToolFixture({ name: 'exampleTool' })]
      expect(mapToolChoiceToOpenAI(tools, 'nonexistentTool')).toBe('none')
    })
  })
})
