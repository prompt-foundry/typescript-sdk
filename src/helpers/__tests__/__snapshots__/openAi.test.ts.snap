// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openAi helpers mapPromptToOpenAIConfig should map the configuration to OpenAI parameters correctly - tools 1`] = `
{
  "frequency_penalty": 0.1,
  "max_tokens": 150,
  "messages": [
    {
      "content": "Hello world",
      "name": undefined,
      "role": "user",
    },
  ],
  "model": "text-davinci-002",
  "presence_penalty": 0.1,
  "response_format": {
    "type": "json_object",
  },
  "seed": 42,
  "temperature": 0.7,
  "tool_choice": "auto",
  "tools": [
    {
      "function": {
        "description": "description",
        "name": "exampleTool",
        "parameters": {},
      },
      "type": "function",
    },
  ],
  "top_p": 0.5,
  "user": undefined,
}
`;

exports[`openAi helpers mapPromptToOpenAIConfig should map the configuration to OpenAI parameters correctly 1`] = `
{
  "frequency_penalty": 0.1,
  "max_tokens": 150,
  "messages": [
    {
      "content": "Hello world",
      "name": undefined,
      "role": "user",
    },
  ],
  "model": "text-davinci-002",
  "presence_penalty": 0.1,
  "response_format": {
    "type": "json_object",
  },
  "seed": 42,
  "temperature": 0.7,
  "tool_choice": undefined,
  "tools": undefined,
  "top_p": 0.5,
  "user": undefined,
}
`;
